mission_systems/navigator_find_the_break/find_the_break_perception.py

```diff
@@ -0,0 +1,206 @@
+import txros
+from twisted.internet import defer
+from txros import util
+from navigator_tools import fprint, CvDebug
+from sensor_msgs.msg import Image
+from cv_bridge import CvBridge
+import numpy as np
+import cv2
+from collections import Counter
+import numpy.ma as ma
+
+# Perception
+# Params : direction
+
+class FindTheBreakPerception(object):
+
+    def __init__(self, nh):
+        """Initialize FindTheBreakPerception class."""
+        # PARAMS
+        self.diff_thresh = 50
+        self.target_wh_ratio = .2
+        self.max_from_target_ratio = .3
+
+        self.nh = nh
+        self.hpipe_position = None
+        self.curr_image = None
+        self.bridge = CvBridge()
+        self.old_hpipe_pos = []
+        self.old_vpipe_pos = []
+        self.hpipe_found = False
+        self.count = 0
+        self.debug = CvDebug(nh=nh)
+
+    @util.cancellableInlineCallbacks
+    def init_(self):
+        """Initialize the txros aspect of FindTheBreakPerception."""
+        self._image_sub = yield self.nh.subscribe("/down/down/image_rect_color", Image, lambda x: x)
+
+    @property
+    @util.cancellableInlineCallbacks
+    def _curr_image(self):
+        img_msg = yield txros.util.wrap_timeout(self._image_sub.get_next_message(), 3)
+        defer.returnValue(self.bridge.imgmsg_to_cv2(img_msg, 'bgr8'))
+
+    def _get_all_pipes(self, frame):
+        frame
+        h, w, r = frame.shape
+        if h == 0:
+            return
+
+        # resize image
+        frame = cv2.resize(frame, 0, fx=.2, fy=.2)
+        gaussian = cv2.GaussianBlur(frame, (9, 9), 10.0)
+        frame = cv2.addWeighted(frame, 1.5, gaussian, -0.5, 0, frame)
+
+        nh, nw, r = frame.shape
+
+        # cluster
+        Z = frame.reshape((-1, 3))
+        Z = np.float32(Z)
+        criteria = (cv2.TERM_CRITERIA_EPS + cv2.TERM_CRITERIA_MAX_ITER, 10, 1.0)
+        K = 5
+        ret, label, center = cv2.kmeans(Z, K, criteria, 10, 0)
+        center = np.uint8(center)
+        image_as_centers = center[label.flatten()]
+        image_as_centers = image_as_centers.reshape((frame.shape))
+        labels = label.reshape((frame.shape[:2]))
+
+        # Get the bounding boxes of the clusters
+        rects = []
+
+        for i in range(0, K):
+            ind = np.where(labels == i)
+            points = ind
+            points = [[points[1], points[0]]]
```

Comment:
No need to do a 2D list here I think, if you just do `np.array([points[1], points[2])` you should be swell.